### DISCUSSION TITLE
Identify direct vs indirect group membership for a person

### DISCUSSION THREAD

**Discussion by mechanarchy** [kanidm/kanidm#3020] [Q&A]
For a person that is `memberOf` a group, is there a way to differentiate between whether that is a `directMemberOf` relationship or a nested/inherited `memberOf` relationship?

The reason for this is that I have added a bunch of accounts and groups to the system, and done manual account&rarr;group membership. Later, I added some group nesting to simplify the system (access is assigned to roles rather than users). However, I now have the situation that this means the `memberOf` relationship is invisibly duplicated: if I need to remove one of the nested groups from a user to remove their access, it won't actually work since the original `directMemberOf` relationship is still in the database.

The closest I've gotten is this kludge on the backing sqlite database, since I couldn't figure out a way to `kanidm raw search` it:
```sql
select rdn.rdn, json_extract(dmo.idl, '$.t.s') as dmo, json_extract(mo.idl, '$.t.s') as mo
from idx_eq_directmemberof dmo
left join idx_eq_memberof mo on mo.key=dmo.key
left join idx_uuid2rdn rdn on rdn.uuid=dmo.key
where dmo.idl <> mo.idl
  and rdn.rdn not like 'spn=idm_%'
  and rdn.rdn not like 'spn=domain_admins@%'
order by dmo.key;
```

Returns something like (edited for clarity):
```
spn=vm_root_user@idm.domain.tld|[253,281]|[253,276,281]
spn=vpn_access@idm.domain.tld|[253,281]|[253,276,281]
spn=infra_admin@idm.domain.tld|[281]|[253,276,281]
```

Here, `253` and `276` are two accounts which are both members of the nested group `281`. Person `253` still has direct membership of the first two groups, so even if I take `253` out of `281` the membership will be retained.

**Comment by yaleman** [comment]
You should be able to query the `directmemberof` attribute for the user and get that list?

**Reply by Firstyear** [reply]
Yes, this is correct.

**Reply by mechanarchy** [reply]
Thanks! Any pointers on how I can do this? I have tried a few different `kanidm raw search` syntaxes but I don't think I have the familiarity yet to work it out.

**Reply by Firstyear** [reply]
What are you trying to achieve and how? Like a service account search of the api? LDAP ? Something else? What's the bigger picture? :)

**Reply by mechanarchy** [reply]
Pretty much as put above - I'd like to identify the groups that a person is a member of both directly and also through an indirect path via nested groups. This is because I was previously assigning permissions to users, rather than roles. Consider:
```
$ kanidm person create person1 person1 --name idm_admin
$ kanidm group create perm_reboot --name idm_admin
$ kanidm group create role_admin --name idm_admin
$ kanidm group add-members perm_reboot role_admin person1 --name idm_admin
$ kanidm group add-members role_admin person1 --name idm_admin
```

Now `person1` is a member of both `perm_reboot` and `role_admin`:
```
$ kanidm person get person1 --name idm_admin
...
memberof: perm_reboot@idm.domain.tld
memberof: role_admin@idm.domain.tld
name: person1
```

Let's assume `person1` changes job description and is no longer an admin. Removing them from the group won't remove their `perm_reboot` permission, because it was previously directly attached:
```
$ kanidm group remove-members role_admin person1 --name idm_admin
$ kanidm person get person1 --name idm_admin
...
memberof: perm_reboot@idm.domain.tld
name: person1
```

So my question is, how can I identify this situation?

**Reply by mechanarchy** [reply]
> Like a service account search of the api? LDAP ? Something else?

For now I'm just using the `kanidm` CLI. On the presumption I can fix all the earlier mistakes and ensure users are only added to 'role' groups rather than 'permission' groups, this should pretty much be a one-time process. Happy to use the API if that's easier but probably sufficient to just make a bash script to fix it once I know the process.

**Reply by Firstyear** [reply]
Actually I think the issue is pretty obvious @yaleman - while we have directmemberof, it's not exposed because we don't have the access controls to allow it today.

I propose either we add directmemberof in addition to any place where memberof is used in access controls for read *OR* we setup an implied search allowed of DMO if MO is allowed in the server code, meaning that we won't forget any locations in acp's. I think I prefer the latter.

**Reply by yaleman** [reply]
I guess the question is, how complicated is adding a special case for the implied search and where's it going to sting us at some point later 😄 Otherwise it seems sensible.

**Reply by Firstyear** [reply]
It's really easy to add, and I doubt it? If you can read memberof, that's the super-set of directmemberof. So how could it sting us (famous last words). Plus it's only in the read path - none of the other paths.

**Reply by yaleman** [reply]
Yeah, just one of those "we add a branch thing at x and ... WHY IS THERE CRABS IN THE FRIDGE" situations.

Dooooeeeeet.
